src/futures/digital.rs

```diff
@@ -0,0 +1,121 @@
+//! Asynchronous digital I/O
+//!
+//! # Examples
+//! ```rust
+//! # use embedded_hal::futures::digital::AsyncInputPin;
+//! //! Asynchronously wait until the `ready_pin` becomes high.
+//! async fn wait_until_ready<P>(ready_pin: &P)
+//! where
+//!     P: WaitFor,
+//! {
+//!     ready_pin
+//!         .wait_for_high()
+//!         .await
+//!         .expect("failed to await input pin")
+//! }
+//! ```
+//!
+//! ```rust,ignore
+//! # use embedded_hal::futures::digital::WaitForHigh;
+//! # use embedded_hal::futures::delay::Delay;
+//! use core::time::Duration;
+//!
+//! //! Wait until the `ready_pin` is high or timeout after 1 millisecond.
+//! //! Returns true if the pin became high or false if it timed-out.
+//! async fn wait_until_ready_or_timeout<P, D>(ready_pin: &P, delay: &mut D) -> bool
+//! where
+//!     P: WaitForHigh,
+//!     D: Delay,
+//! {
+//!     futures::select_biased! {
+//!         x => ready_pin.wait_for_high() => {
+//!             x.expect("failed to await input pin");
+//!             true
+//!         },
+//!         _ => delay.delay(Duration::from_millis(1)) => false, // ignore the error
+//!     }
+//! }
+//! ```
+
+use core::future::Future;
+
+/// Asynchronously wait for a pin to be high.
+pub trait WaitForHigh {
+    /// Enumeration of errors.
+    type Error;
+
+    /// The future returned by the `wait_for_high` function.
+    type Fut<'a>: Future<Output = Result<(), Self::Error>> + 'a
```

Comment:
Naming the associated future type `Fut` will cause naming conflicts if a type `Pin` impls both `WaitForHigh` and `WaitForLow` and the user tries to do `Pin::Fut`.

It's a very niche edge case, since users will usually `.await` the futures directly instead of storing them somewhere, but still.

I think the best convention is: for a method named `foo_bar`, name the associated future `FooBarFuture`. All other traits follow this convention, imo these here should too

Reply:
Yeah, I guess this makes sense. It seems so verbose to list out the entire method in the associated type, but it avoids naming conflicts as you said.